libvast/vast/hash_index.hpp

```diff
@@ -0,0 +1,213 @@
+
+  /// Computes a chopped digest from arbitrary data.
+  /// @param x The data to hash.
+  /// @param seed The seed to use during the hash.
+  /// @returns The chopped digest.
+  static digest_type hash(data_view x, size_t seed = 0) {
+    digest_type result;
+    auto digest = uhash<hasher_type>{seed}(x);
+    std::memcpy(result.data(), &digest, Bytes);
+    return result;
+  }
+
+  /// Constructs a hash index for a particular type and digest cutoff.
+  /// @param t The type associated with this index.
+  /// @param digest_bytes The number of bytes to keep of a hash digest.
+  explicit hash_index(vast::type t) : value_index{std::move(t)} {
+  }
+
+  caf::error serialize(caf::serializer& sink) const override {
+    // Prune unneeded seeds.
+    decltype(seeds_) non_null_seeds;
+    for (auto& [k, v] : seeds_)
+      if (v > 0)
+        non_null_seeds.emplace(k, v);
+    return caf::error::eval([&] { return value_index::serialize(sink); },
+                            [&] { return sink(digests_, non_null_seeds); });
+  }
+
+  caf::error deserialize(caf::deserializer& source) override {
+    return caf::error::eval([&] { return value_index::deserialize(source); },
+                            [&] { return source(digests_, seeds_); });
+  }
+
+private:
+  struct key {
+    friend bool operator==(key x, digest_type y) {
+      return std::memcmp(x.bytes.data(), y.data(), y.size()) == 0;
+    }
+
+    friend bool operator!=(key x, digest_type y) {
+      return !(x == y);
+    }
+
+    friend bool operator==(digest_type x, key y) {
+      return y == x;
+    }
+
+    friend bool operator!=(digest_type x, key y) {
+      return !(x == y);
+    }
+
+    friend bool operator==(key x, key y) {
+      return x == y.bytes;
+    }
+
+    friend bool operator!=(key x, key y) {
+      return !(x == y);
+    }
+
+    digest_type bytes;
+  };
+
+  struct key_hasher {
+    auto operator()(const key& x) const {
+      auto result = uint64_t{0};
+      std::memcpy(&result, x.bytes.data(), x.bytes.size());
+      return result;
+    }
+  };
+
+  // Retrieves the unique digest for a given input or generates a new one.
+  caf::optional<key> make_digest(data_view x) {
+    size_t i = 0;
+    do {
+      // Compute a hash digest.
+      auto digest = hash(x, i);
+      auto k = key{digest};
+      // If we have never seen this digest before, we're adding it to the list
+      // of seen digests and are done.
+      if (unique_digests_.count(k) == 0) {
+        auto materialized = materialize(x);
+        VAST_ASSERT(seeds_.count(materialized) == 0);
+        seeds_.emplace(std::move(materialized), i);
+        unique_digests_.insert(k);
+        return k;
+      };
+      // If we have seen the digest, check whether we also have a known
+      // preimage.
+      // FIXME: do not materialize every time, but this requires the ability to
+      // use key equivalence (C++20).
+      if (auto it = seeds_.find(materialize(x)); it != seeds_.end())
```

Comment:
@dominiklohmann is there a way to have C++20 key equivalence already in C++17? I don't want to materialize a `data_view` every time, but simply search based on the view instance.

Reply:
For C++17, set the template parameter for the unordered map manually. For C++20, the `KeyEqual` template param is set to `Hash::transparent_key_equal` instead iff it is available, with `Hash` being another template parameter.

The default values are this:

```c++
class Hash = std::hash<Key>,
class KeyEqual = std::equal_to<Key>>
```

This means that all you need to do is set a custom equality function here:

```c++
// define custom comparator somewhere
auto comp = [](const data& lhs, const data& rhs) -> bool { ... }

// change declaration of seeds map
std::unordered_map<data, size_t, std::hash<data>, decltype(comp)> seeds_;
```

Reply:
I'm not sure how the custom comparator will solve this issue. The map contains `data`, but I'd like `find` to work with `data_view`, without having to materialize the views to a full-blown `data` instance.

Reply:
Oh I see. So your issue really is that `find` is hardcoded to use the same `Key` as the map instead of some arbitrary type that can be equality compared with `KeyEqual` with the `Key` being on the lhs. You can't get that right now.

A dirty hack I can imagine to work is having the Key be a union of `data` and `data_view`, and using `data_view` on the rhs of your custom `KeyEqual` function and `data` everywhere else.

Reply:
`boost::unordered_map` has this:

```c++
template<typename CompatibleKey,
         typename CompatibleHash, 
         typename CompatiblePredicate> 
iterator find(CompatibleKey const&,
              CompatibleHash const&, 
              CompatiblePredicate const&);
```

Maybe we should just copy this over and add a `TODO:` comment for switching back to `std::unordered_map` once we arrive at C++20.

Reply:
> Maybe we should just copy this over

Copying over the entire implementation of a boost header? All the dependencies, the MPL, etc. will cause too much bloat for the desired gain. I think we have to figure out a different solution. For example, we could switch to `flat_map` for which we have our own implementation.

Reply:
```c++
template <typename Key, typename T, typename Hash, typename KeyEqual,
          typename Allocator>
auto find_with_precomputed_hash(
    const std::unordered_map<Key, T, Hash, KeyEqual, Allocator> &data,
    std::invoke_result_t<Hash, Key> precomputed_hash) {
  return std::find_if(data.begin(), data.end(),
                      [precomputed_hash = std::move(precomputed_hash),
                       hash_fn = data.hash_function()](const auto &lhs) {
                        return hash_fn(lhs.first) == precomputed_hash;
                      });
}
```

This should work for your purpose just fine.

Reply:
More general approach that should work for any unordered container in the standard library:

```c++
#include <algorithm>
#include <type_traits>

template <typename PrecomputedHash,
          template <typename...> class UnorderedContainer, typename... Ts>
auto find_with_precomputed_hash(UnorderedContainer<Ts...>& container,
                                PrecomputedHash&& precomputed_hash)
  -> decltype(
    std::declval<
      std::invoke_result_t<typename UnorderedContainer<Ts...>::hasher,
                           typename UnorderedContainer<Ts...>::key_type>>()
      == std::declval<PrecomputedHash>(),
    container.begin()) {
  return std::find_if(container.begin(), container.end(),
                      [precomputed_hash
                       = std::forward<PrecomputedHash>(precomputed_hash),
                       hash_fn = container.hash_function()](auto& lhs) {
                        return hash_fn(lhs.first) == precomputed_hash;
                      });
}
```

I actually think this is okay-ish in terms of readability.

See test at: https://godbolt.org/z/jQwQaU

Reply:
Isn't this O(n) due to `find_if`?

Reply:
If you're planning to not have any collisions then the `std::unordered_map<...>::find(...)` should not be any better than O(end-begin) either, right?